Allow command options to be set in .ember-cli file

kpfefferle opened the pull request:
This is a crude first attempt at using a setting within a project's `.ember-cli` file to get verbose deploy outputs on every deploy by default. It is used by simply adding `"deployVerbose": true` to your project's `.ember-cli` file.

The implementation is inspired by ember-cli's own `usePods` setting (code: https://github.com/ember-cli/ember-cli/blob/3c615aa55dfe6be1adf14ef67e74028b09560138/lib/commands/destroy.js#L68-L70) so that the `--verbose` flag can still be used with `ember deploy` to reverse the default.

Potential areas for feedback:
- [x] Do we like the setting name `deployVerbose`? I believe this could even be nested as `"deploy": { "verbose": true }` if we want to go down the road of putting more of these settings (such as `activateOnDeploy`) in the `.ember-cli` file.
- [x] Is there a style or readability preference between reversing the value of `this.ui.verbose` after it's assigned (as implemented), conditionally assigning the value `this.ui.verbose`, or using a one-line ternary assignment?
- [ ] Testing: how and where should this be tested? I don't see any comparable tests in this project that would be easily adapted to test the input from `.ember-cli`. Feedback or direction here would be especially welcomed.

ghedamat commented:
@kpfefferle thanks for this!

I'm not a fan of using the flag `--verbose` to reverse the behaviour of the settings in `.ember-cli` but if that's already the convention I'm ok with it.

I'm also :+1: to nest the keys in case we later need to add options and keep things isolated.

for the testing part, I'd assume we'd have to stub the config file, maybe worth looking into how ember-cli does that?

lukemelia commented:
Thanks @kpfefferle. Cool to see how easy this actually is. I agree with with @ghedamat re: nesting keys, perhaps even under `ember-cli-deploy`?

I'm strongly against `--verbose` ever meaning the opposite of verbose. I'd prefer to override via `--verbose=false` or `--no-verbose`.

Re: testing, the sad fact is that we don't have any unit test coverage at the command level. I would really like us to get there. It doesn't have to this PR though.

kpfefferle commented:
@ghedamat @lukemelia I made a couple of updates in response to your feedback:
- The `.ember-cli` setting is now formatted as `"ember-cli-deploy": { "verbose": true }`
- This setting sets the default, but can be overridden by an explicitly passed flag i.e. `ember deploy production --verbose=false`

lukemelia commented on lib/commands/deploy.js:
```diff
@@ -15,7 +15,15 @@ module.exports = {
+    } else {
+      this.ui.verbose = commandOptions.verbose;
+    }
+
```
Maybe clearer as:

```js
this.ui.verbose = commandOptions.verbose;
if (this.ui.verbose === undefined) {
  this.ui.verbose = this.settings && this.settings['ember-cli-deploy'] && this.settings['ember-cli-deploy'].verbose;
}
```

kpfefferle replied:
@lukemelia Your phrasing would change the code's behavior slightly in that it assumes that the `.ember-cli` setting is `true` rather than passing through the setting value (which I suppose would be ok as long as the default remains `false` and the setting is only to override it).

I was attempting to check for (1) the existence of the setting value and (2) the absence of an explicit flag before (3) assigning the actual value of the setting. If you prefer the initial assignment to occur before the conditional, maybe this instead?
```javascript
this.ui.verbose = commandOptions.verbose;
if (this.ui.verbose === undefined && this.settings && this.settings['ember-cli-deploy'] && this.settings['ember-cli-deploy'].verbose) {
  this.ui.verbose = this.settings['ember-cli-deploy'].verbose;
}
```
It's a long conditional, but it retains my original functionality.

lukemelia replied:
@kpfefferle :+1:

kpfefferle commented:
@lukemelia Updated per your feedback!

lukemelia commented:
OK, cool. Thanks @kpfefferle. I'm thinking we should also add support for the `--activate` and `--config` options (am I missing any others?) and update the commit message with something the we can paste into the 0.6.x docs when we create them.

kpfefferle commented:
New commits:
- Copied the reading of the `--verbose` setting from `.ember-cli` to the `deploy:list` and `deploy:activate` commands
- Added similar support for `"ember-cli-deploy": { "activate": true }` to the `deploy` command

Clarification is needed before I add support for the `--deploy-config-file` and `--amount` (for `deploy:list`) flags. These flags [have defaults set in the `availableOptions` settings](https://github.com/ember-cli/ember-cli-deploy/blob/master/lib/commands/list.js#L10-L14) which complicate knowing whether the default value has come from this code or whether the user explicitly set a value that matches the default. I believe that the priority order should be as follows:

1. Explicitly set value (i.e. `--deploy-config-file=config/deploy.js`) always takes top priority.
1. Setting defined in `.ember-cli` takes secondary priority.
1. Default is only used if the above two are not defined.

To accomplish this, I will need to remove the default values from `availableOptions` so that I can check and see if these flags have been explicitly set or not. Then I can accurately apply the `.ember-cli` setting or default value in the correct priority order.

Does this seem like the proper approach to you @lukemelia?

lukemelia commented:
@kpfefferle yep, that seems reasonable to me. My only question is whether removing the defaults will negatively effect `ember help` output, and whether we can remedy that by including the default in the description

kpfefferle commented:
@lukemelia I think I've got that all covered now. I added a helper function to choose between the passed value, the `.ember-cli` value, and the default value to clean things up a bit. I also added descriptions to the command options that explain the defaults using the same text that the CLI was generating before.

Do you prefer that all of this get squashed into one commit with a documentation-ready message?

lukemelia commented:
Yes, looks good! Please do squash @kpfefferle. Thanks for your work on this!

On Tuesday, December 29, 2015, Kevin Pfefferle <notifications@github.com>
wrote:

> @lukemelia <https://github.com/lukemelia> I think I've got that all
> covered now. I added a helper function to choose between the passed value,
> the .ember-cli value, and the default value to clean things up a bit. I
> also added descriptions to the command options that explain the defaults
> using the same text that the CLI was generating before.
>
> Do you prefer that all of this get squashed into one commit with a
> documentation-ready message?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ember-cli/ember-cli-deploy/pull/331#issuecomment-167898446>
> .
>

ghedamat commented:
LGTM (aside from squashing) @kpfefferle thanks!

kpfefferle commented:
@lukemelia @ghedamat: Squashed! Thanks for your input!
